### ISSUE TITLE
num_hits flag in run_mturk.py may be incorrect

### ISSUE THREAD

**Issue by taneemtee** [facebookresearch/ParlAI#89]
Hello from Mechanical Turk! I am part of the MTurk team, and I'm trying to walk through the MTurk integration in ParlAI.

In the "run_mturk.py" file, one of the parameters is "num_hits", which can be set to choose how many HITs are published for a specific task. I think there may be a logical bug here. 

E.g. if I set num_hits = 3 while using model_evaluator: 

1. ParlAI publishes 3 HITs to MTurk containing the details of 1 task
2. The same Worker could do all three HITs (doing one HIT does not stop you from doing another HIT), rating the same task over and over again
3. The net result is that you have the same result (a score from ONE Worker for ONE task) repeated 3 times - so it's less information than you wanted and you're paying the same person 3 times. 

I believe what you really want is 3 _different_ Workers rating each task. MTurk supports this using the concept of _assignments_. What you want is to publish 1 HIT per Task, but with 3 assignments per HIT. You can define the number of assignments using the "MaxAssignments" flag when creating HITs (http://docs.aws.amazon.com/AWSMechTurk/latest/AWSMturkAPI/ApiReference_CreateHITOperation.html).

If you ask for 3 assignments for a HIT, MTurk will make the HIT available to 3 different Workers automatically. A Worker is not allowed to complete more than one assignment per HIT.

Then, when you are retrieving results, you will want to check for all 3 assignments being done, instead of searching for all 3 HITs getting done.

To recap: right now for each task, ParlAI posts X HITs with 1 Assignments each, and should instead post 1 HIT with X Assignments. 

Please let me know if that makes sense and if I can help in any other way.

**Comment by yf225** [comment]
Hi Taneem,

Thanks for your interest! Here are my explanations to our design choices:

1. Currently, we don't really care about whether it's one Worker completing all of the jobs, or unique Workers that work on each of them. Enforcing unique Worker per job would certainly encourage response diversity, but it could also increase the overall completion time (since we will have to wait for other workers to look at the jobs). We would consider adding "unique worker" as an option in the future if there is enough interest in the ParlAI community.

2. If we set num_hits = 3, ParlAI will publish 3 HITs to MTurk containing the details of 1 task. However, the content for each of the HITs is provided by the local agent and thus can be different, depending on how the user implements the act() method of the local agent. In other words, the 3 HITs for this task only share the structural layout of the task, but not the content itself.

The only reason the model_evaluator example shows three identical dialogs for evaluation is that we don't support randomized test set for MovieDD-Reddit dataset at the moment, and thus the first dialog from the test set is always used. As a comparison, if you run the qa_data_collection example, each of HIT will contain a different Wikipedia paragraph, which is expected.

Hope this is helpful for understanding the codebase. We are planning to add it to our upcoming tutorial for MTurk. Please let me know if I can help in any other way.

**Comment by taneemtee** [comment]
Thanks for the context, that definitely makes sense. I hadn't fully grokked point #2 while poking around and the explanation now makes sense. I am preparing a blog post outlining how to use ParlAI, but if you're planning to publish a tutorial soon I may wait for that and use it as a reference instead. 

I'm closing this issue as I think there's nothing else needed here. Thanks again!

**Comment by yf225** [comment]
We will have the MTurk tutorial ready this week or next. Thanks for your suggestions too!

**Comment by yf225** [comment]
@taneemtee We've published a tutorial for using MTurk [here](http://parl.ai/static/docs/mturk.html). Please let me know if I can be of any help for the blog post. Thanks!

**Comment by taneemtee** [comment]
@yf225 Thanks for sharing this! We just added a [blog post](https://blog.mturk.com/tutorial-training-ai-bots-to-chat-using-mturk-and-facebooks-parlai-42af81475c6c) pointing to the tutorial, with some additional help for users who are new to MTurk (feedback welcome!). This gets added to our [collection of tutorials](https://blog.mturk.com/tutorials/home), so it'll persist over time. We'll also post a note to our Twitter account at the start of the work week next Tuesday.

**Comment by yf225** [comment]
@taneemtee Thanks so much for the blog post! The steps for setting up AWS and MTurk accounts are really helpful and we want to add them to our tutorial too - I am wondering is there a way to create a bookmark link to the "Connect to MTurk" section so that I can put a reference to it from our tutorial? Also for the ParlAI announcement link at the beginning we can use [the FAIR blog post](https://code.facebook.com/posts/266433647155520/parlai-a-new-software-platform-for-dialog-research/) which contains much more information about the project as well. Thanks again!

**Comment by taneemtee** [comment]
Ok great - I've updated the link to point to the FAIR post. To my knowledge there is no way to create anchor links in Medium unfortunately. However, please feel free to add as much of the content as you like from our blog post directly into your guide.

**Comment by yf225** [comment]
Awesome I added the MTurk registration steps into our guide. Thanks!

**Comment by yf225** [comment]
@taneemtee We've pushed a redesign to our MTurk code and now the steps for running an MTurk task is a bit different - now the user would simply need to go into one of the task directories (e.g. ``parlai/mturk/tasks/model_evaluator``), and then run ``python run.py -nh <num_hits> -r <reward> [--sandbox]/[--live]``, with the params set accordingly. Do you mind updating the blog post to reflect this change? Thanks and really appreciate it.

**Comment by taneemtee** [comment]
Thanks for the heads up - I've just made some updates - can you kindly take a look to see if the changes are correctly reflected?

**Comment by yf225** [comment]
Thanks a lot for the update! There are two places that I found might need to be changed: one is that in each task folder we will have worlds.py and run.py instead of agents.py now, and the other is that we won't use the `verbose` flag anymore and instead let the user decide what messages to output to terminal by leaving the implementation to them. 

I created a sample of the changes needed which could be helpful:

Here you will find folders representing tasks that you can perform using ParlAI with MTurk - collecting data (**qa_data_collection**) and evaluating your dialog models (**model_evaluator**). Each task folder contains an worlds.py and a task_config.py file. The worlds.py file allows you to create a world where the dialog model and MTurk Workers will interact in. Both tasks come configured with defaults that we can use to test that we have ParlAI setup properly. The “task_config.py” allows you to override various settings like the title and description that ParlAI uses when creating HITs on MTurk.

Each task folder also has a “run.py” file which is what you will use to run that specific task.

Let’s try running an example model_evaluator task, publish two HITs (tasks for Workers), paying $0.05 for each but in the MTurk Sandbox for testing only. From the tasks/model_evaluator folder, run:

`$ python run.py -nh 2 -r 0.05 --sandbox`

The first time you run this, ParlAI will download the dataset that has been specified in the run.py file. By default it will use the #MovieDD-Reddit dataset which is a collection of dialog from the Movies subreddit on Reddit.
